content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md

```diff
@@ -0,0 +1,152 @@
+package io.reflectoring.nullsafety;
+
+import org.springframework.lang.NonNullFields;
+```
+
+Now, **we no longer need to annotate the fields with `@NonNull`**. Because by default, all the fields are now treated as non-null. Thanks to `@NonNullFields`! And, we will still see the same warning as before:
+
+{{% image alt="IDE warning for NonNullFields" src="images/posts/spring-boot-null-safety-annotations/nonnull-ide-warning.png" %}}
+
+Another point to note here is if there are any uninitialized fields, then we will see a warning to initialize those:
+
+{{% image alt="IDE warning for NonNull" src="images/posts/spring-boot-null-safety-annotations/nonnullfields-ide-warning.png" %}}
+
+## `@NonNullApi`
+
+By now, you might have spotted another requirement, i.e., to have similar checks for method parameters or return values. Here [`@NonNullApi`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullApi.html) will come to our rescue. 
+
+In the same fashion, as we did for `@NonNullFields`, we can use a `package-info.java` file and add the `@NonNullApi` annotation for the intended package:
+
+```java
+@NonNullApi
+package io.reflectoring.nullsafety;
+
+import org.springframework.lang.NonNullApi;
+```
+
+We can now see the IDE is warning us about the non-nullable parameters:
+
+{{% image alt="IDE warning for NonNullApi" src="images/posts/spring-boot-null-safety-annotations/nonnullapi-ide-warning.png" %}}
+
+## `@Nullable`
+
+But here is a catch. **There could be scenarios where a particular field can be null** (no matter how much we want to avoid!). 
+
+For example, the `pastEmployment` field could be nullable in the `Employee` class (for a fresher!). But as per our safety checks, the IDE thinks it cannot be.
+
+Fear not! We can simply express our intention using the `@Nullable` field. This will tell the IDE that the field can be null in some cases, so no need to trigger an alarm. As the [JavaDoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/Nullable.html) suggests,
+> Can be used in association with `@NonNullApi` or `@NonNullFields` to override the default non-nullable semantic to nullable.
+
+Similar to `NonNull`, the `Nullable` annotation can be applied to the method, parameter, or field level. 
+
+We can now mark the `pastEmployment` field as nullable:
+
+```java
+package io.reflectoring.nullsafety;
+
+\\ imports
+
+class Employee {
+
+  @Nullable
+  String pastEmployment;
+
+  @Nullable String getPastEmployment() {
+    return pastEmployment;
+  }
+
+  \\...
+```
+
+## Conclusion
+
+These annotations are indeed a boon for the Java programmers to reduce the possibility of a `NullPointerException` arising during runtime. But, please bear in mind this does not guarantee complete null safety.
+
+**However, please note that not all development tools can show these warnings. If you don't see the compilation warning, check the compiler settings in IntelliJ.**
```

Comment:
Also, is there a way to evaluate these annotations during a Maven or Gradle build? That would be interesting...and certainly worth another section in this article :)

Reply:
I found the [SpotBugs](https://spotbugs.readthedocs.io/en/latest/introduction.html) Maven/Gradle Plugin is able to offer some help. Added a section on that.